### PR TITLE
GH-35870: [C++] Add support for changing optimization flags with CMAKE_CXX_FLAGS_DEBUG

### LINKED ISSUE #35870: [C++] Cannot change optimization level using ARROW_CXXFLAGS
### Describe the bug, including details regarding any error messages, version, and platform.

`ARROW_CXXFLAGS` used to allow changing the optimization level. However, this now gets ignored as the default optimization flags get appended _after_ the `ARROW_CXXFLAGS`.

For example, the `conda-cpp-valgrind` job produces the following CMake configuration:
```
-- CMAKE_C_FLAGS: -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/arrow/include -fuse-ld=gold  -Wall -fno-semantic-interposition -msse4.2 -Og
-- CMAKE_CXX_FLAGS:  -Wno-noexcept-type -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/arrow/include -fdiagnostics-color=always -fuse-ld=gold  -Wall -fno-semantic-interposition -msse4.2 -Og
-- CMAKE_C_FLAGS_DEBUG: -g -O0 -ggdb
-- CMAKE_CXX_FLAGS_DEBUG: -g -O0 -ggdb
```

... and then the following compiler commands:
```
[1/684] /opt/conda/envs/arrow/bin/ccache /opt/conda/envs/arrow/bin/x86_64-conda-linux-gnu-c++ -DARROW_EXPORTING -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_HDFS -DARROW_VALGRIND -DARROW_WITH_BACKTRACE -DARROW_WITH_BROTLI -DARROW_WITH_BZ2 -DARROW_WITH_LZ4 -DARROW_WITH_RE2 -DARROW_WITH_SNAPPY -DARROW_WITH_TIMING_TESTS -DARROW_WITH_UTF8PROC -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DURI_STATIC_BUILD -I/build/cpp/src -I/arrow/cpp/src -I/arrow/cpp/src/generated -isystem /arrow/cpp/thirdparty/flatbuffers/include -isystem /arrow/cpp/thirdparty/hadoop/include -Wno-noexcept-type -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/arrow/include -fdiagnostics-color=always -fuse-ld=gold  -Wall -fno-semantic-interposition -msse4.2 -Og -g -O0 -ggdb -std=c++17 -fPIC -pthread -MD -MT src/arrow/CMakeFiles/arrow_objlib.dir/util/bitmap_builders.cc.o -MF src/arrow/CMakeFiles/arrow_objlib.dir/util/bitmap_builders.cc.o.d -o src/arrow/CMakeFiles/arrow_objlib.dir/util/bitmap_builders.cc.o -c /arrow/cpp/src/arrow/util/bitmap_builders.cc
```
(notice how `-O0` overrides `-Og` by coming after it)

`ARROW_CXXFLAGS` should ideally override any default settings.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

`-DCMAKE_CXX_FLAGS_RELEASE=-O...` works with #15022 but `-DCMAKE_CXX_FLAGS_DEBUG=-O...` and
`-DCMAKE_CXX_FLAGS_RELWITHDEBINFO=-O...` don't work. Because we always specify `-O0` for Debug and RelWithDebInfo.

### What changes are included in this PR?

These changes don't specify optimization levels for Debug and RelWithDebInfo unconditionally. We don't specify optimization levels for RelWithDebInfo the default flags include them. We specify `-O0` for Debug only when the default flags don't include any optimization levels. So user specified `CMAKE_CXX_FLAGS_DEBUG=-O...` and `CMAKE_CXX_FLAGS_RELWITHDEBINFO=-O...` are used.

In addition, `ARROW_C_FLAGS_${CONFIG}` and `ARROW_CXX_FLAGS_${CONFIG}` are introduced. Users can also use `-DARROW_CXX_FLAGS_DEBUG=-O...` instead of `-DCMAKE_CXX_FLAGS_DEBUG=-O...`.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* Closes: #35870

### CODE REVIEW

**Comment by kou** [comment]
@ursabot please benchmark

**Comment by kou** [comment]
@ursabot please benchmark

**Comment by austin3dickey** [comment]
Runs are also viewable at https://buildkite.com/apache-arrow -- we're making some improvements to the benchmarking notifications!

**Review by pitrou — APPROVED** [review]
I tried `ARROW_CXX_FLAGS_RELWITHDEBINFO` and it seems to work here. Do we want to document compiler option settings somewhere?

**Comment by kou** [comment]
Good point. I've added it.

File: docs/source/developers/cpp/building.rst [hunk 1]
```diff
@@ -254,6 +254,34 @@ Several build types are possible:
...
+   ``-DARROW_CXX_FLAGS_RELWITHDEBINFO=...``) for the ``RelWithDebInfo``
+   build type.
+
+   For example, you can use ``-O3`` as an optimization flag by
+   specifying ``-DARROW_CXX_FLAGS_RELEASE=-O3`` for the ``Release``
+   build type. You can use ``-g3`` as a debug flag by specifying
+   ``-DARROW_CXX_FLAGS_DEBUG=-g3`` for the ``Debug`` build type.
+
+   You can also use the standard ``-DCMAKE_C_FLAGS_${BUILD_TYPE}=...``
+   and ``-DCMAKE_CXX_FLAGS_${BUILD_TYPE}=...`` style but
```
Comment (pitrou) [inline docs/source/developers/cpp/building.rst:276]:
```suggestion
   and ``-DCMAKE_CXX_FLAGS_${BUILD_TYPE}=...`` variables, but
```

File: docs/source/developers/cpp/building.rst [hunk 2]
```diff
@@ -254,6 +254,34 @@ Several build types are possible:
...
+   ``-DARROW_C_FLAGS_${BUILD_TYPE`` and
+   ``-DARROW_CXX_FLAGS_${BUILD_TYPE}=...`` style is
+   recommended. Because ``-DCMAKE_C_FLAGS_${BUILD_TYPE}=...`` and
+   ``-DCMAKE_CXX_FLAGS_${BUILD_TYPE}=...`` style replaces all default
+   flags provided by CMake. ``-DARROW_C_FLAGS_${BUILD_TYPE`` and
+   ``-DARROW_CXX_FLAGS_${BUILD_TYPE}=...`` style just overrides the
+   default flags specified instead of replacing them.
```
Comment (pitrou) [inline docs/source/developers/cpp/building.rst:283]:
```suggestion
   the ``ARROW_C_FLAGS_${BUILD_TYPE}`` and
   ``ARROW_CXX_FLAGS_${BUILD_TYPE}`` variables are
   recommended. The ``CMAKE_C_FLAGS_${BUILD_TYPE}`` and
   ``CMAKE_CXX_FLAGS_${BUILD_TYPE}`` variables replace all default
   flags provided by CMake, while ``ARROW_C_FLAGS_${BUILD_TYPE}`` and
   ``ARROW_CXX_FLAGS_${BUILD_TYPE}`` just append the
   flags specified, which allows selectively overriding some of the defaults.
```

File: docs/source/developers/cpp/building.rst [hunk 3]
```diff
@@ -254,6 +254,34 @@ Several build types are possible:
...
+   (``-DARROW_C_FLAGS_DEBUG=...`` / ``-DARROW_CXX_FLAGS_DEBUG=...``) for the
+   ``Debug`` build type and ``RELWITHDEBINFO``
+   (``-DARROW_C_FLAGS_RELWITHDEBINFO=...`` /
+   ``-DARROW_CXX_FLAGS_RELWITHDEBINFO=...``) for the ``RelWithDebInfo``
+   build type.
+
+   For example, you can use ``-O3`` as an optimization flag by
+   specifying ``-DARROW_CXX_FLAGS_RELEASE=-O3`` for the ``Release``
+   build type. You can use ``-g3`` as a debug flag by specifying
+   ``-DARROW_CXX_FLAGS_DEBUG=-g3`` for the ``Debug`` build type.
```
Comment (pitrou) [inline docs/source/developers/cpp/building.rst:273]:
```suggestion
   For example, you can use ``-O3`` as an optimization flag for the ``Release``
   build type by passing ``-DARROW_CXX_FLAGS_RELEASE=-O3`` .
   You can use ``-g3`` as a debug flag for the ``Debug`` build type
   by passing ``-DARROW_CXX_FLAGS_DEBUG=-g3`` .
```

**Review by pitrou — APPROVED** [review]

**Comment by kou** [comment]
Thanks for these suggestions!

**Pull request merged**